packages/block-editor/src/store/actions.js

```diff
@@ -157,18 +158,96 @@ export function receiveBlocks( blocks ) {
  * @param {boolean}         uniqueByBlock true if each block in clientIds array has a unique set of attributes
  * @return {Object} Action object.
  */
-export function updateBlockAttributes(
-	clientIds,
-	attributes,
-	uniqueByBlock = false
-) {
-	return {
-		type: 'UPDATE_BLOCK_ATTRIBUTES',
-		clientIds: castArray( clientIds ),
-		attributes,
-		uniqueByBlock,
+export const updateBlockAttributes =
+	( clientIds, attributes, uniqueByBlock = false ) =>
+	( { select, dispatch, registry } ) => {
+		if ( ! window.__experimentalPatterns ) {
+			dispatch( {
+				type: 'UPDATE_BLOCK_ATTRIBUTES',
+				clientIds: castArray( clientIds ),
+				attributes,
+				uniqueByBlock,
+			} );
+			return;
+		}
+
+		const updates = {};
+		for ( const clientId of castArray( clientIds ) ) {
+			const attrs = uniqueByBlock ? attributes[ clientId ] : attributes;
+			const parentBlocks = select.getBlocksByClientId(
+				select.getBlockParents( clientId )
+			);
+			const parentPattern = parentBlocks.findLast(
+				( parentBlock ) => parentBlock.name === 'core/block'
+			);
+			const block = select.getBlock( clientId );
+			if (
+				! parentPattern ||
+				! registry
+					.select( blocksStore )
+					.hasBlockSupport( block.name, '__experimentalPattern' )
+			) {
+				updates[ clientId ] = attrs;
+				continue;
+			}
```

Comment:
I like the explorations you and @aaronrobertshaw have worked on!

Given the block bindings API is going to support a bunch of different data sources, I wonder how this can be made generic/pluggable, and done in a way where the `core/block` isn't hard-coded in.

Having read through the [block bindings issue](https://github.com/WordPress/gutenberg/issues/54536), it seems like the blocks inside the pattern will have some config denoting their connections (which I guess the pattern creator will define), which could be read here. Then there's the need to discover the source of the data for those connections, which is what happens here.

I think maybe we're close to the point where we can try modelling those connections and abstracting the discovery of the data source that happens here. It could be that the connection 'type' determines the strategy used to find the data source (the pattern block). Or alternatively the pattern block is defined as source.

But we should at the same time connect back with the other data bindings work around custom fields to make sure the approach taken works in both places. Looking at the POCs so far for custom fields (https://github.com/WordPress/gutenberg/issues/53300), the challenge is that they focus more on the PHP part as it stands.

Reply:
Yeah, exactly which blocks and their attributes should be connected was still undecided. This part was just following the existing exploration in the base branch.

The main goal for this PR is to find a performant and extendible way to allow override of `setAttributes`. Writing condition paths directly in `updateBlockAttributes` works for prototyping, but I'm currently trying to find a better approach so that it feels like "duct-taping". I've tried maybe 6 different approaches now but each has its trade-offs and challenges. Once we solve that, and we have the connection "contract" defined, then I think we'll be in a good shape for implementing them.